### PR TITLE
Fetch firmware versions for Cmis Target FW upgrade supported cables

### PR DESCRIPTION
**Pull request by mihirpat1**
#### Description

With onboarding of CMIS target firmware upgrade supported transceivers, we need a mechanism to store the firmware version of E1 and E2 targets (in addition to E0 target) in redis-db. The current change set is for accomplishing this task.
XCVRD will be able to push the target firmware version information to TRANSCEIVER_FIRMWARE_INFO table of STATE_DB.

MSFT ADO - 26565884

#### Motivation and Context

This change is required to periodically read target firmware versions and store it in TRANSCEIVER_FIRMWARE_INFO table of STATE_DB.

#### How Has This Been Tested?

Following scenarios were tested

1.1	CLI output on transceiver not supporting Target FW upgrade
1.2	XCVR on target 1 removed
1.3	XCVR on target 2 removed
1.4	All targets having different inactive firmware versions
1.5	CLI output on transceiver not supporting Target FW upgrade
1.6	CLI on port with DOM disabled

### CODE REVIEW

**Comment by prgeor** [comment]
@mihirpat1 type in PR description -> "E1 and E1 targets "

File: sonic_platform_base/sonic_xcvr/api/xcvr_api.py [hunk 1]
```diff
@@ -75,8 +75,8 @@ def get_transceiver_info_firmware_versions(self):
         Retrieves active and inactive firmware versions of the xcvr
 
         Returns:
-            A list with active and inactive firmware versions of the xcvr
-            [active_firmware, inactive_firmware]
+            A dictionary containing the active and inactive firmware versions of the transceiver
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/xcvr_api.py:78]: @AnoopKamath please note this change for your PR...

File: sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py [hunk 2]
```diff
@@ -5,9 +5,139 @@
...
+
+TARGET_LIST = [TARGET_E0_VALUE, TARGET_E1_VALUE, TARGET_E2_VALUE]
+
+AEC_E1_FIRMWARE_INFO_MAP = {
+    'active_firmware': 'e1_active_firmware',
+    'inactive_firmware': 'e1_inactive_firmware',
+    'server_firmware': 'e1_server_firmware'
+}
+
+AEC_E2_FIRMWARE_INFO_MAP = {
```
[the remark this answers is not in the document]
Comment (mihirpat1) [reply sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py]: @prgeor I have addressed this now.

File: sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py [hunk 3]
```diff
@@ -5,9 +5,139 @@
...
+
+            if calculated_checksum & 0xFF == checksum:
+                server_fw_version_str = ''
+                for i in range(0, SERVER_FW_VERSION_SIZE, SERVER_FW_VERSION_NUMBER_SIZE):
+                    server_fw_version_number = bytes(server_fw_version_byte_array[i:i+SERVER_FW_VERSION_NUMBER_SIZE])
+                    # Each number of the server firmware version is 4 bytes and is stored in big endian format.
+                    # Convert the 4 bytes to a number and then convert the number to a string.
+                    server_fw_version_str += str(struct.unpack('>I', server_fw_version_number)[0]) + '.'
+                server_fw_version_str = server_fw_version_str[:-1]
+                return_dict['server_firmware'] = server_fw_version_str
```
[the remark this answers is not in the document]
Comment (mihirpat1) [reply sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py]: @prgeor I have addressed this now.

**Review by prgeor — APPROVED** [review]

**Comment by mihirpat1** [comment]
@StormLiangMS @yxieca Can you please help in merging this to 202305 and 202311?
MSFT ADO - 26565884

File: sonic_platform_base/sonic_xcvr/api/xcvr_api.py (hunk 1 above)
Comment (AnoopKamath) [reply sonic_platform_base/sonic_xcvr/api/xcvr_api.py:78]: Ack. Thanks @prgeor

**Pull request merged**